specifications/SimplifiedFastPaxos/FastPaxos.tla

```diff
@@ -0,0 +1,170 @@
+*)
+FastPropose ==
+    /\ UNCHANGED<<decision, cValue>>
+    /\ \E a \in Replicas, m \in p2aMessages, v \in Values:
+        /\ m.value = any
+        /\ maxBallot[a] <= m.ballot
+        /\ maxValue[a] = none \/ maxValue[a] = v
+        /\ maxBallot' = [maxBallot EXCEPT ![a] = m.ballot]
+        /\ maxVBallot' = [maxVBallot EXCEPT ![a] = m.ballot]
+        /\ maxValue' = [maxValue EXCEPT ![a] = v]
+        /\ \A n \in p2bMessages : ~(n.ballot = m.ballot /\ n.acceptor = a)
+        /\ SendMessage([type |-> "P2b",
+                        ballot |-> m.ballot,
+                        acceptor |-> a,
+                        value |-> v])
+
+(*
+    A value is chosen if a fast quorum of acceptors proposed that value in a fast round.
+
+    Because the quorum size of a fast round and classic round is different, we assume that the acceptor distinguishes
+    a fast round and classic round based on the P2a message it receives. If the P2a message contains the special value
+    "any", it is a fast round. Else it is a classic round.
+*)
+FastDecide ==
+    /\ UNCHANGED<<messages, maxBallot, maxVBallot, maxValue, cValue>>
+    /\ \E b \in FastBallots, q \in FastQuorums :
+        LET M == {m \in p2bMessages : m.ballot = b /\ m.acceptor \in q}
+            V == {w \in Values : \E m \in M : w = m.value}
+        IN /\ \A a \in q : \E m \in M : m.acceptor = a
+           /\ 1 = Cardinality(V)
+           /\ \E m \in M : decision' = m.value
+
+(*
+    Phase 2a (Classic)
+
+    If more than one value has been proposed, the collision is resolved using the following rules:
+
+    1. If the proposals contain different values, a value must be selected if the majority of
+       acceptors in the fast quorum have casted a vote for that value.
+
+    2. Otherwise, the coordinator is free to select any value.
+*)
+ClassicAccept ==
+    /\ UNCHANGED<<decision, maxBallot, maxVBallot, maxValue>>
+    /\ \E b \in ClassicBallots, f \in FastBallots, q \in FastQuorums, v \in Values :
+        /\ f < b \* There was a fast round before this classic round.
+        /\ cValue = none \/ cValue = v
+        /\ cValue' = v
+        /\ \A m \in p2aMessages : m.ballot # b
+        /\ LET M == {m \in p2bMessages : m.ballot = f /\ m.acceptor \in q}
+               V == {w \in Values : \E m \in M : w = m.value}
+           IN /\ \A a \in q : \E m \in M : m.acceptor = a
+              /\ 1 < Cardinality(V) \* Collision occured.
+              /\ IF \E w \in V : IsMajorityValue(M, w)
+                 THEN IsMajorityValue(M, v) \* Choose majority in quorum.
+                 ELSE v \in V \* Choose any.
+              /\ SendMessage([type |-> "P2a",
+                              ballot |-> b,
+                              value |-> v])
+
+(*
+    Phase 2b (Classic)
+
+    Same as in Paxos.
+*)
+ClassicAccepted ==
+    /\ UNCHANGED<<cValue>>
+    /\ PaxosAccepted
+
+(*
+    Same as in Paxos.
+*)
+ClassicDecide ==
```

Comment:
It would have been a little nicer if this definition had been written as
/\ UNCHANGED cValue
/\ PaxosDecide
which, I believe, is equivalent, but makes the correspondence more explicit.